### ISSUE TITLE
Boolean conditionals forced to break regardless of print width inside JSX expression

### ISSUE THREAD

**Issue by adamroyle** [prettier/prettier#2483]
Not sure if this is intentional? I couldn't find the logic in the source code to figure out where this logic is applied.

Before prettier...

```jsx
<div>
  {cond1 && cond2 && (
    <Element
      options={that}
      force={the}
      element={to}
      break={over}
      multiple={lines}
    />
  )}
</div>
```
After prettier...
```jsx
<div>
  {cond1 &&
  cond2 && (
    <Element
      options={that}
      force={the}
      element={to}
      break={over}
      multiple={lines}
    />
  )}
</div>
```

Can we detect this pattern and only break the conditionals if the exceed print width?

**Comment by azz** [comment]
Prettier probably thinks that `cond1`, `cond2` and `<Element />` are in the same group, which does exceed the print width.

**Comment by azz** [comment]
Looks like:
```js
let x = cond1 || cond2 || (
    {
      a
    }
)
```
Formats to:
```js
let x = cond1 ||
cond2 || {
  a
};
```

**Comment by vjeux** [comment]
It seems like we are missing one level of indentation there. I wouldn't want cond1 && cond2 to be inlined.

**Comment by jackyho112** [comment]
@vjeux 

Hello, I would like to contribute to `prettier` and this seems like a good first issue. Should this be the correct output?

```
let x = cond1 ||
  cond2 || {
    a
  };
```

Any other pointer is greatly appreciated! 😄

**Comment by vjeux** [comment]
@jackyho112 sorry I was in vacation. That is indeed the correct expected indentation. If you look at `printBinaryish` function, there is some piece of code that looks at the parent and figure out if it should be indented or not. Looks like we're not doing the right thing here.

**Comment by jackyho112** [comment]
@vjeux 
Thank you so much for replying and the tip! I was looking at the file where`printBinaryish` is defined as well. I love what `prettier` is doing and would love to tackle this issue hopefully starting next week at the latest. 

I will keep you posted and if I have any question, I will leave a message here.

**Comment by jackyho112** [comment]
@vjeux, @azz 

Sorry about the delay. I have been busy lately, but I took a stab at solving the bug yesterday night and did not have much luck. Fortunately, I did learn quite a lot about the part of the library related to this issue.

In the functions, `printBinaryishExpressions` and `genericPrintNoParens`, I am not sure if there is a way to check the parent node or/and current node so that only the specific cases of this issue get indented without breaking a bunch of tests and knowing how the previous parts are printed. 

I think maybe the problem is that once you enter a new line for the object literal and `doc.break` becomes `true` for the whole group and its nested groups, every `line` is a new line whether the expression fits on the line or not and for the first `line` in that group, there also needs to be an `indent` for the format to look right?

With that said, maybe this is an acceptable format?

```
let x = cond1 || cond2 || {
  a
};
```

Sorry if my interpretation is incorrect or confusing. Let me know if you have any clue.

This is not fixed,

```
let x = cond1 || cond2 || (
    {
      a
    }
)
```

but this is fixed now.

```
<div>
  {cond1 && cond2 && (
    <Element
      options={that}
      force={the}
      element={to}
      break={over}
      multiple={lines}
    />
  )}
</div>
```

is now formatted to

```
<div>
  {cond1 &&
    cond2 &&
    <Element
      options={that}
      force={the}
      element={to}
      break={over}
      multiple={lines}
    />}
</div>;
```
